encoder_decoder_test.go

```diff
@@ -0,0 +1,716 @@
+			Input:  []any{func() {}, "supported", make(chan struct{}), "supported"},
+			Output: typeTree{_type: wafArrayType, children: []typeTree{{_type: wafStringType}, {_type: wafStringType}}},
+		},
+		{
+			Name:   "unsupported-array-values",
+			Input:  []any{func() {}, "supported", make(chan struct{}), "supported"},
+			Output: typeTree{_type: wafArrayType, children: []typeTree{{_type: wafStringType}, {_type: wafStringType}}},
+		},
+		{
+			Name:   "unsupported-array-values",
+			Input:  []any{func() {}, make(chan struct{}), "supported"},
+			Output: typeTree{_type: wafArrayType, children: []typeTree{{_type: wafStringType}}},
+		},
+		{
+			Name: "unsupported-map-key-types",
+			Input: map[any]int{
+				"supported":           1,
+				interface{ m() }(nil): 1,
+				nil:                   1,
+				(*int)(nil):           1,
+				(*string)(nil):        1,
+				make(chan struct{}):   1,
+			},
+			Output: typeTree{_type: wafMapType, children: []typeTree{{_type: wafIntType}}},
+		},
+		{
+			Name: "unsupported-map-key-types",
+			Input: map[any]int{
+				interface{ m() }(nil): 1,
+				nil:                   1,
+				(*int)(nil):           1,
+				(*string)(nil):        1,
+				make(chan struct{}):   1,
+			},
+			Output: typeTree{_type: wafMapType},
+		},
+		{
+			Name: "unsupported-map-values",
+			Input: map[string]any{
+				"k0": func() {},
+				"k2": make(chan struct{}),
+			},
+			Output: typeTree{_type: wafMapType, children: []typeTree{{_type: wafInvalidType}, {_type: wafInvalidType}}},
+		},
+		{
+			Name: "nil-map-values",
+			Input: map[string]any{
+				"k0": (*struct{ a int })(nil),
+				"k1": (*string)(nil),
+				"k2": (*int)(nil),
+			},
+			Output: typeTree{_type: wafMapType, children: []typeTree{{_type: wafNilType}, {_type: wafNilType}, {_type: wafNilType}}},
+		},
+	} {
+		encoder := newMaxEncoder()
+		encoded, err := encoder.Encode(tc.Input)
+		t.Run(tc.Name+"/assert", func(t *testing.T) {
+			if tc.Error != nil {
+				require.Error(t, err, "expected an encoding error when encoding %v", tc.Input)
+				require.Equal(t, tc.Error, err)
+				return
+			}
+
+			require.NoError(t, err, "unexpected error when encoding: %v", err)
+			assertEqualType(t, tc.Output, encoded)
+		})
+
+		t.Run(tc.Name+"/run", func(t *testing.T) {
+			wafTest(t, encoded)
+			keepAlive(encoder.cgoRefs)
+		})
+	}
+}
+
+// This test needs a working encoder to function properly, as it first encodes the objects before decoding them
+func TestDecoder(t *testing.T) {
+	t.Run("Errors", func(t *testing.T) {
+		e := newMaxEncoder()
+		objBuilder := func(value any) *wafObject {
+			encoded, err := e.Encode(value)
+			require.NoError(t, err, "Encoding object failed")
+			return encoded
+		}
+
+		for _, tc := range []struct {
+			Name          string
+			ExpectedValue map[string][]string
+		}{
+			{
+				Name:          "empty",
+				ExpectedValue: map[string][]string{},
+			},
+			{
+				Name: "one-empty-entry",
+				ExpectedValue: map[string][]string{
+					"afasdfafs": nil,
```

Comment:
love the 🛼 ⌨️

Reply:
This one is not even from me 🤣 it's from you

https://github.com/DataDog/go-libddwaf/blob/main/decoder_test.go